### PR TITLE
Ft update results

### PR DESCRIPTION
**Pull request by jamsgra**
#### What does this PR do?
- Checks user privilege for updating a result
- Validates PATCH data
- Allows authorised users to update result info with valid PATCH data
- Changes primary key of `Results` table to `result_uid` instead of `id`

#### Description of Task to be completed?
- `checkResultUpdatePrivilege`
- `validateUpdateResults`
- `updateResult`
- Tests

#### How should this be manually tested?

- `npm i`
- `npm test`

#### Any background context you want to provide?

- N/A

### CODE REVIEW

File: server/routes/teacherPrivilege.js [hunk 1]
```diff
@@ -17,6 +17,9 @@ router.post('/api/v1/users/update', checkIfUserHasSchool, checkUserUpdatePrivile
 // adds new result to the database
 router.post('/api/v1/results/addresults', checkIfUserHasSchool, AddResultsValidation.validateAddResults, ResultsController.addResults);
 
+// update result in the database
+router.post('/api/v1/results/update', checkIfUserHasSchool, checkResultUpdatePrivilege, AddResultsValidation.validateUpdateResults, ResultsController.updateResult);
```
Comment (sulenchy) [inline server/routes/teacherPrivilege.js]: @jamsgra I do not think the appropriate HTTP verb to be used here is `post`. According to the resource at https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods, the appropriate HTTP verb should be either PUT or PATCH.

Comment (jamsgra) [reply server/routes/teacherPrivilege.js]: @sulenchy you are right, I have pushed changes to use PUT for both `updateUser` & `updateResult`

Comment (jamsgra) [reply server/routes/teacherPrivilege.js]: @sulenchy Actually looking at it, PATCH may be the more suitable option as the entire record isn't necessarily updated. What do you think?

File: server/controllers/resultsController.js [hunk 2]
```diff
@@ -140,6 +140,29 @@ class ResultsController {
       return sendError(res, 500, err);
     }
   }
+
+  static async updateResult(req, res) {
+    try {
+      const updateData = res.locals.result;
```
Comment (sulenchy) [inline server/controllers/resultsController.js]: @jamsgra Is there any reason you prefer `res.locals` to `req.body`?

File: server/controllers/resultsController.js [hunk 3]
```diff
@@ -140,6 +140,29 @@ class ResultsController {
...
+      const updatedResult = await Results.update(updateData, {
+        where: {
+          result_uid: req.body.result_uid
```
Comment (sulenchy) [inline server/controllers/resultsController.js]: @jamsgra I think we should not be having the result_uid in the req.body object? we should be having it in the req.query or req.params. What do you think?

Comment (jamsgra) [reply server/controllers/resultsController.js]: I agree, I've moved `result_uid` & `user_uid` to `req.query`

File: server/middlewares/addResultsValidation.js [hunk 4]
```diff
@@ -70,7 +88,7 @@ class addResultsValidation {
             email: ExcelValidators.validateEmail(value),
             year: ExcelValidators.validateYear(value),
             term: ExcelValidators.validateTerm(value),
-            subject: ExcelValidators.isAlphanumeric(value),
+            subject: ExcelValidators.isAlpha(value),
```
Comment (sulenchy) [inline server/middlewares/addResultsValidation.js:91]: Nice catch here

**Review by sulenchy — APPROVED** [review]
Look good to me

**Pull request merged**